src/components/ui/Chat/MessageList/styleMessageList.tsx

```diff
@@ -0,0 +1,111 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React from 'react';
+
+import DateHeader from './DateHeader';
+import ChatBubble from '../ChatBubble';
+import { PopOverItemProps } from '../../PopOver/PopOverItem';
+
+export interface MessageProps {
+  /* The displayed text of the message sent. */
+  content: string;
+  /* The unique identifier of the message. */
+  messageId: string;
+  /* The timestamp when the message was originally sent by the sender. */
+  createdTimestamp: string;
+  /* The timestamp of the last time the message was edited. */
+  lastUpdatedTimestamp: string;
+  /* Determines if a message was redacted (deleted) by a user. */
+  redacted: boolean;
+  /* The unique identifier of the channel the message was sent in. */
+  channelArn?: string;
+  sender: {
+    /* The display name of the sender. */
+    name: string;
+    /* The unique identifier of the sender. */
+    arn: string;
+  }
+};
+
+export const styleMessageList = (messages: MessageProps[], memberArn: string, messageActions: PopOverItemProps[] | null) => {
+
+  const formatTime = (time: string) => {
+    const t = new Date(time).toLocaleTimeString();
+    return t.slice(0, t.length - 6) + t.slice(t.length - 3, t.length);
+  }
+
+  const insertDateHeaders = (messageItems: MessageProps[]) => {
+    const items: any[] = [...messageItems]
+    const dateMap = {};
+    let messageDate = messageItems[0].createdTimestamp;
+    let dateCount = 0;
+    messageItems.forEach((m: any, i: number) => {
+      if (!m.content) {
+        return // not a message
+      }
+      if (i === 0) {
+        items.splice(0, 0, <DateHeader key={i.toString()} date={m.createdTimestamp} />)
+        dateMap[new Date(m.createdTimestamp).toLocaleDateString()] = 1
+        dateCount++
+      } else if (
+          new Date(m.createdTimestamp).toLocaleDateString() !== messageDate &&
+          !dateMap[new Date(m.createdTimestamp).toLocaleDateString()]
+        ) {  
+        items.splice(i + dateCount, 0, <DateHeader key={i.toString()} date={m.createdTimestamp} />)
+        messageDate = new Date(m.createdTimestamp).toLocaleDateString()
+        dateMap[new Date(m.createdTimestamp).toLocaleDateString()] = 1
+        dateCount++
+      }
+    })
+    return items
+  }
+
+  const createChatBubbles = (items: MessageProps[]) => {
+    return items.map((m: any, i: number) => {
+      if (!m.content) {
+        return m // not a message
+      } else {
+        let showTail = true;
+        if (
+          items[i].sender?.arn && // it is a message
+          items[i + 1]?.sender?.arn && // the item after is a message
+          items[i].sender.arn === items[i + 1].sender.arn // the item after is from the same sender
```

Comment:
You sometimes have optional chaining here, and sometimes not. I'd recommend choosing the most defensive option. Also, you're dipping into the `sender` object each time for the `arn`. It would be more performant to cache both of those values.

Reply:
I don't have optional chaining where I have already checked the values. If `items[i].sender` is falsey it would fail out on line 71. I can cache the values early, though. I'll update.